group/secretsharing/ss.go

```
@@ -0,0 +1,138 @@
+// Feldman's verifiable secret sharing: https://ieeexplore.ieee.org/document/4568297
+package secretsharing
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/cloudflare/circl/group"
+)
+
+type SecretShare struct {
+	ID    uint64
+	Share group.Scalar
+}
+
+// SecretSharing implements Shamir's secret sharing.
+type SecretSharing interface {
+	Shard(rnd io.Reader, secret group.Scalar) []SecretShare
+	Recover(shares []SecretShare) (secret group.Scalar, err error)
+}
+
+type secretSharing struct {
+	G    group.Group
+	T, N uint64
+}
+
+// New returns a struct implementing SecretSharing interface.
+func New(g group.Group, t, n uint64) (secretSharing, error) {
+	if !(0 < t && t <= n) || g == nil {
+		return secretSharing{}, errors.New("secretsharing: bad parameters")
+	}
+	ss := secretSharing{G: g, T: t, N: n}
+	var _ SecretSharing = ss // checking at compile-time
+	return ss, nil
+}
+
+func (s secretSharing) polyFromSecret(rnd io.Reader, secret group.Scalar) (p polynomial) {
+	p = randomPolynomial(rnd, s.G, s.T)
+	p.coeff[0] = secret.Copy()
+	return
+}
+
+func (s secretSharing) generateShares(poly polynomial) []SecretShare {
+	shares := make([]SecretShare, s.N)
+	x := s.G.NewScalar()
+	for i := range shares {
+		id := uint64(i + 1)
+		x.SetUint64(id)
+		shares[i].ID = id
+		shares[i].Share = poly.evaluate(x)
+	}
+
+	return shares
+}
+
+func (s secretSharing) Shard(rnd io.Reader, secret group.Scalar) []SecretShare {
+	return s.generateShares(s.polyFromSecret(rnd, secret))
+}
+
+func (s secretSharing) Recover(shares []SecretShare) (group.Scalar, error) {
+	if l := len(shares); l <= int(s.T) {
+		return nil, fmt.Errorf("secretsharing: do not met threshold %v with %v shares", s.T, l)
+	} else if l > int(s.N) {
+		return nil, fmt.Errorf("secretsharing: %v shares above max number of shares %v", l, s.N)
+	}
+
+	x := make([]group.Scalar, len(shares))
+	px := make([]group.Scalar, len(shares))
+	for i := range shares {
+		x[i] = s.G.NewScalar()
+		x[i].SetUint64(shares[i].ID)
```

Passing the same share multiple times leads to silent failure (well, depending on whether the scalar errors on inverting zero.)

---

now, the Lagrange module checks whether the shares are all different.